### PR TITLE
Add AWS Elastic Container Service authentication support

### PR DESCRIPTION
**Pull request by toelen**

### CODE REVIEW

**Comment by ben-gineer** [comment]
It would be great to see this in the docker client

File: src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    if (accessToken == null) {
+      // has not yet been fetched
+      return true;
+    }
+
+    final Date expirationTime = accessToken.getExpiresAt();
+
+    // Don't refresh if expiration time hasn't been provided.
+    if (expirationTime == null) {
+      return true;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java]: Do we need to test this? Will `accessToken.getExpiresAt()` ever be null?

File: src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  private AmazonECR ecrClient = mock(AmazonECR.class);
+
+  private final AuthorizationData authData1 = new AuthorizationData()
+      .withAuthorizationToken(tokenValue).withExpiresAt(expiration.toDate())
+      .withProxyEndpoint("http://proxy");
+
+  private final Clock clock = mock(Clock.class);
+  private final int minimumExpirationSecs = 30;
+  private final EcrCredentials credentials = spy(new EcrCredentials(ecrClient, authData1));
+  private final ContainerRegistryAuthSupplier supplier = spy(new ContainerRegistryAuthSupplier(
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java]: I don't think you need `spy()` here. See https://gist.github.com/davidxia/ad4b35ae0777c30726afabe0a77490c0

**Comment by mhermosi** [comment]
Any news on this MR? any time frame for its availability?

**Comment by smuryginim** [comment]
Hey team. Let merge this PR?

**Comment by davidxia** [comment]
This PR doesn't seem ready yet to merge. I at least still had some comments and unresolved questions above.

**Comment by toelen** [comment]
@davidxia I fixed your remarks on https://github.com/toelen/docker-client/tree/feat/ecr-auth

**Comment by mattnworb** [comment]
Sorry for the late reply.

> In general, I'm not sure if we want to add more vendor specific libraries to this repo or if we want to pull out this support to separate libraries that can be plugged in.
>
> See spotify/dockerfile-maven#54.
> 
> @mattnworb thoughts?

I think the best thing would be to structure these things as different modules, meaning change docker-client to be a multi-module maven build with a "core" module that contains all the existing code minus the GCR part, which can go in a separate module, and then ECR or other RegistryAuthSuppliers can all go in different modules. But I am indifferent about if that is a prerequisite for merging this change or not.

File: src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    final long expiresIn = expirationTime.getTime() - clock.currentTimeMillis();
+
+    return expiresIn <= minimumExpiryMillis;
+  }
+
+  @Override
+  public RegistryAuth authFor(final String imageName) throws DockerException {
+    final String[] imageParts = imageName.split("/", 2);
+
+    if ((imageParts.length < 2) || !imageParts[0].contains(".dkr.ecr.")
```
Comment (acmcelwee) [inline src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java]: What about just doing a regex on the ECR repo format? Something along the lines of `^\d+\.dkr\.ecr\.[a-z-0-9]+\.amazonaws\.com\/`

File: src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        Matchers.any(GetAuthorizationTokenRequest.class));
+  }
+
+  @Test
+  public void testAuthForImage_NoRefresh() throws Exception {
+    when(clock.currentTimeMillis()).thenReturn(
+        expiration.minusSeconds(minimumExpirationSecs + 1).getMillis());
+
+    assertThat(
+        supplier.authFor("1234567890.dkr.ecr.eu-west-1.amazonaws.com/foobar/barfoo:latest"),
```
Comment (acmcelwee) [inline src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java]: All of these image names seem to be the wrong format, to me. Shouldn't it be `1234567890.dkr.ecr.eu-west-1.amazonaws.com/foobar:latest`?

File: src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+    verify(credentials, never()).refresh();
+  }
+
+  @Test
+  public void testAuthForImage_RefreshNeeded() throws Exception {
+    doReturn(true).when(supplier).needsRefresh(Matchers.any(AuthorizationData.class));
+
+    assertThat(
+        supplier.authFor("1234567890.dkr.ecr.eu-west-1.amazonaws.com/foobar/barfoo:latest"),
```
Comment (acmcelwee) [inline src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java]: Same image name format comment as above

File: src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+    verify(credentials).refresh();
+  }
+
+  @Test
+  public void testAuthForImage_TokenExpired() throws Exception {
+    doReturn(true).when(supplier).needsRefresh(Matchers.any(AuthorizationData.class));
+
+    assertThat(
+        supplier.authFor("1234567890.dkr.ecr.eu-west-1.amazonaws.com/foobar/barfoo:latest"),
```
Comment (acmcelwee) [inline src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java]: Same image name format comment as above

File: src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    doReturn(true).when(supplier).needsRefresh(Matchers.any(AuthorizationData.class));
+
+    final IOException ex = new IOException("failure!!");
+    doThrow(ex).when(credentials).refresh();
+
+    // the exception should propagate up
+    exception.expect(DockerException.class);
+    exception.expectCause(is(ex));
+
+    supplier.authFor("1234567890.dkr.ecr.eu-west-1.amazonaws.com/example/foobar:1.2.3");
```
Comment (acmcelwee) [inline src/test/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplierTest.java]: Same image name format comment as above

File: src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java (hunk 1 above)
Comment (acmcelwee) [reply src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java]: It's technically nullable - https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-ecr/src/main/java/com/amazonaws/services/ecr/model/AuthorizationData.java

**Comment by sigpwned** [comment]
Oops! I just submitted a different pull request for ECR authentication. I missed this in my search. I was searching for ECR and ECS. Hopefully now this comment will flag for future searchers.

It looks like this review may have stalled out. If I can do some work to push it through, or if my other pull request looks OK, it would be great to get ECR support into this project soon. Very happy to help do that — just LMK how!

File: src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+  // TODO (mbrown): change to java.time.Clock once on Java 8
+  private final Clock clock;
+  private EcrCredentials credentials;
+  private final long minimumExpiryMillis;
+
+  @VisibleForTesting
+  ContainerRegistryAuthSupplier(final AmazonECR ecr, final Clock clock,
+                                final long minimumExpiryMillis, final EcrCredentials credentials) {
+    Preconditions.checkArgument(ecr != null, "ecr");
+    this.ecr = ecr;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java]: Do we need to pass `ecr` here? I see it's used as a sync object below, but can't you just lock on something else?

**Comment by davidxia** [comment]
@toelen Sorry, I didn't see your [comment here](https://github.com/spotify/docker-client/pull/876#issuecomment-369854873).

> @davidxia I fixed your remarks on https://github.com/toelen/docker-client/tree/feat/ecr-auth

Can you update this PR with those changes?

File: src/main/java/com/spotify/docker/client/auth/ecr/EcrCredentials.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  EcrCredentials(final AmazonECR ecr, final AuthorizationData authorizationData) {
+    this.ecr = ecr;
+    this.authorizationData = authorizationData;
+  }
+
+  public AuthorizationData getAuthorizationData() {
+    return authorizationData;
+  }
+
+  void refresh() throws IOException {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/auth/ecr/EcrCredentials.java]: You don't need the IOException here.

File: src/main/java/com/spotify/docker/client/auth/ecr/EcrCredentials.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  EcrCredentials(final AmazonECR ecr) {
+    this(ecr, null);
+  }
+
+  EcrCredentials(final AmazonECR ecr, final AuthorizationData authorizationData) {
+    this.ecr = ecr;
+    this.authorizationData = authorizationData;
+  }
+
+  public AuthorizationData getAuthorizationData() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/auth/ecr/EcrCredentials.java]: Can be package-private.

File: src/main/java/com/spotify/docker/client/auth/ecr/EcrCredentials.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    return authorizationData;
+  }
+
+  void refresh() throws IOException {
+    GetAuthorizationTokenResult authorizationToken = ecr
+        .getAuthorizationToken(new GetAuthorizationTokenRequest());
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/auth/ecr/EcrCredentials.java]: javadocs for `GetAuthorizationTokenRequest()` say

> A list of AWS account IDs that are associated with the registries for which to get authorization tokens. If you do not specify a registry, the default registry is assumed.

Will the user ever need to use registries other than the default?

**Comment by cenacsf** [comment]
Can We get this merge please , We also eagerly waiting for this support

**Comment by davidxia** [comment]
@cenacsf There's this and https://github.com/spotify/docker-client/pull/1013. The last update was my comment here https://github.com/spotify/docker-client/pull/1013#issuecomment-392994030. It seems like both of these PRs have stalled. If someone's willing to drive https://github.com/spotify/docker-client/pull/1013#issuecomment-392994030 to completion, I'm happy to review and merge.

**Pull request closed without merging**